### PR TITLE
fix: Fixes for https://github.com/antlr/antlr4/issues/3718

### PR DESCRIPTION
**Pull request by jimidle**
fix: #3718 Implement generic collections and fiux LA(*) optimizer

  o Implements collections using generics to solve has conflicts in previous implementation of
     DFAState collections and alt conflict resolution
  o Change alt and config to use correct hash and equals as per Java
  o Fix LA()*) runtime optimization routine
  o A few tidy ups on comments etc

NB: There are now extra hash() and equals() routines that I had to leave there unless I wanted to go through
all the code and rationalize and refactor. Not worth it if we are going to rewrite. The generic collections work for this
purpose but the make up of the currents structs etc, forces a few hacks that won't be needed with better structure.

Sending this PR without tests for the collections as:

  o they can't be worse than what is there, that doesn't work
  o they are tested with real grammars anyway
  o it is Saturday and time to get ready for EPL to start in England - come on Leeds

### CODE REVIEW

**Comment by jimidle** [comment]
I will have to remember the DCO. I think some of the builds are failing because the tests are failing. Somebody changed a test somewhere it seems.

**Comment by kaby76** [comment]
Please check that this fixes the [original SimpleBoolean.g4](https://github.com/kaby76/issue-3718/blob/main/original-grammar/go/SimpleBoolean.g4), [input p15.txt](https://github.com/kaby76/issue-3718/blob/main/p15.txt) performance. I don't see the perf improvement. [My code does](https://github.com/kaby76/issue-3718/tree/main/mod-runtime/go). Strange, your code looks like it hits all the important places, but something is off.

File: runtime/Java/src/org/antlr/v4/runtime/atn/ATNConfigSet.java [hunk 1]
```diff
@@ -74,10 +74,7 @@ public boolean equals(ATNConfig a, ATNConfig b) {
 	 * All configs but hashed by (s, i, _, pi) not including context. Wiped out
 	 * when we go readonly as this set becomes a DFA state.
 	 */
-	public AbstractConfigHashSet configLookup;
-
-	/** Track the elements as they are added to the set; supports get(i) */
-	public final ArrayList<ATNConfig> configs = new ArrayList<ATNConfig>(7);
+	public AbstractConfigHashSet configLo
```
Comment (parrt) [inline runtime/Java/src/org/antlr/v4/runtime/atn/ATNConfigSet.java:77]: Looks like a mistaken chop chop here :)

**Comment by parrt** [comment]
I pull in, fix the java file booboo, but get Go compile errors:

```
critter:~/antlr/code/antlr4/runtime-testsuite $ mvn -Dtest=go.** test
...
# antlr
/var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/ANTLR-runtime-testsuite-cache/Go/src/antlr/atn_config.go:23:27: syntax error: unexpected [, expecting comma or )
/var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/ANTLR-runtime-testsuite-cache/Go/src/antlr/atn_config.go:140:33: syntax error: unexpected [, expecting )
/var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/ANTLR-runtime-testsuite-cache/Go/src/antlr/atn_config.go:145:46: syntax error: unexpected [, expecting comma or )
/var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/ANTLR-runtime-testsuite-cache/Go/src/antlr/atn_config.go:152:2: syntax error: non-declaration statement outside function body
/var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/ANTLR-runtime-testsuite-cache/Go/src/antlr/atn_config.go:158:2: syntax error: non-declaration statement outside function body
/var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/ANTLR-runtime-testsuite-cache/Go/src/antlr/atn_config.go:171:2: syntax error: non-declaration statement outside function body
/var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/ANTLR-runtime-testsuite-cache/Go/src/antlr/atn_config.go:270:37: syntax error: unexpected [, expecting )
/var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/ANTLR-runtime-testsuite-cache/Go/src/antlr/atn_config.go:273:51: syntax error: unexpected [, expecting comma or )
/var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/ANTLR-runtime-testsuite-cache/Go/src/antlr/atn_config.go:276:2: syntax error: non-declaration statement outside function body
/var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/ANTLR-runtime-testsuite-cache/Go/src/antlr/atn_config.go:286:2: syntax error: non-declaration statement outside function body
/var/folders/w1/_nr4stn13lq0rvjdkwh7q8cc0000gn/T/ANTLR-runtime-testsuite-cache/Go/src/antlr/atn_config.go:286:2: too many errors
```

**Comment by parrt** [comment]
I get 5 failures running the tests:

```
critter:~/antlr/code/antlr4/runtime-testsuite $ mvn -Dtest=go.** test
...
```

**Comment by parrt** [comment]
@jimidle I changed CI for hosted github actions to 1.19 Go.

**Comment by jimidle** [comment]
Sorry for the Java file. It’s the kind of thing that happens in a mono
repo. Obviously a keyboard slip in IntelliJ while single stepping there and
in goland at the same time.

I normally don’t submit so many code changes at once.

**Comment by parrt** [comment]
No prob. I do crazy stuff like that all the time. Can you check on the go test failures?

**Comment by jimidle** [comment]
Will do. Hopefully it’s the tests that are incorrect

**Comment by parrt** [comment]
Possibly but we have a general test mechanism that gens tests across targets. You can check the antlr doc dir and see testing file

**Comment by parrt** [comment]
The checks dialog here should also give the results though I just tuned off all but github hosted CI and circleci

**Comment by jimidle** [comment]
> Please check that this fixes the [original SimpleBoolean.g4](https://github.com/kaby76/issue-3718/blob/main/original-grammar/go/SimpleBoolean.g4), [input p15.txt](https://github.com/kaby76/issue-3718/blob/main/p15.txt) performance. I don't see the perf improvement. Looks like it should.

It does, but you need to make sure you are actualling compiling with this code. You cannot do a go get with my fork because of the way go get works. You. have to clone my repo, then either use GOWORK, which is my preference, to tell the compiler to use that code for your build, or use a replace in the go mod for the test code. Or just wait for Ter to merge the PR and use:

```sh
go get -u github.com/antlr/antlr4/runtime/Go/antlr@dev
```

**Comment by jimidle** [comment]
Hmmm - seems to be a syntax error in the push - not sure how that happened...

**Comment by jimidle** [comment]
Ah - the ci go testing the runtime is using a go compiler that is too old. That test also needs updating to use go 1.18 or go 1.19 (preferable).

**Comment by jimidle** [comment]
> @jimidle I changed CI for hosted github actions to 1.19 Go.

Also needs changing for the ci test of the go runtime using the test suite. That is failing because the compiler is too old

**Comment by kaby76** [comment]
This isn't fixing the Go performance problem for me.
* I am cleaning out my ~/go cache, wherever it may be on Ubuntu 20.04.4 LTS and Windows 10.
* I am using go 1.17.2 on both Ubuntu and Windows
* I am using "go get github.com/antlr/antlr4/runtime/Go/antlr@dev".
* I am using the executable on both Linux and Windows.
* I am doing a fresh "git clone https://github.com/kaby76/issue-3718; cd issue-3718/original-grammar/".
* Here is my complete work, all the sources and makefiles that I used to build and test the code, and the output. https://github.com/kaby76/issue-3718/tree/c3ba6e60e18f0bbba19b5113fd5f07b71e9cb869/original-grammar
* Here's the output from "test.sh 2>&1 1> out-win.txt" on Windows. https://github.com/kaby76/issue-3718/blob/c3ba6e60e18f0bbba19b5113fd5f07b71e9cb869/original-grammar/out-win.txt
* Here's the output from "test.sh 2>&1 1> out-linux.txt" on WSL Ubuntu https://github.com/kaby76/issue-3718/blob/c3ba6e60e18f0bbba19b5113fd5f07b71e9cb869/original-grammar/out-linux.txt
* The runtimes (Windows) are: CSharp [0.05s](https://github.com/kaby76/issue-3718/blob/c3ba6e60e18f0bbba19b5113fd5f07b71e9cb869/original-grammar/out-win.txt#L39), Java [0.017s](https://github.com/kaby76/issue-3718/blob/c3ba6e60e18f0bbba19b5113fd5f07b71e9cb869/original-grammar/out-win.txt#L52), Go [38s](https://github.com/kaby76/issue-3718/blob/c3ba6e60e18f0bbba19b5113fd5f07b71e9cb869/original-grammar/out-win.txt#L103) . Similar result for Ubuntu.

Someone else, please verify.

**Comment by jimidle** [comment]
Ken

That isn’t the correct runtime. Wait for Ter to merge the PR

**Comment by kaby76** [comment]
You're right. It does fix the problem. I upgraded to 1.18.3, "learned" about workspaces, and tried it out.  I modeled it after [this](https://go.dev/doc/tutorial/workspaces). It does fix the perf problem.

(I do not understand yet how a go build finds all the workspace code because go.work can in theory have "use" to Go modules anywhere on the entire file system. But, those modules do not have pointers back to the go.work file to pick up all the modified modules in the workspace. If it's relative, then what happens if I copy a stray go.work into "~"?)

**Comment by parrt** [comment]
> Also needs changing for the ci test of the go runtime using the test suite. That is failing because the compiler is too old

I'm not sure where are you mean by ci as the dev branch shows for hosted CI:

```
    - name: Setup Go 1.19
      if: matrix.target == 'go'
      uses: actions/setup-go@v3
      with:
        go-version: '^1.19'
```

**Comment by parrt** [comment]
I think master works but dev doesn't for go on circlei, though we might be able to dump circleci now that github hosted CI works.

Thanks for all the hard work, @jimidle !!!

**Comment by jimidle** [comment]
> I think master works but dev doesn't for go on circlei, though we might be able to dump circleci now that github hosted CI works.
> 
> Thanks for all the hard work, @jimidle !!!

Yes - sorry I was not sure which build test were running where. You are correct that it was the circleci that was trying to use the old compiler and hence that trigger failed. 

I will delete this branch once I verify that the dev branch is in sync

**Comment by jimidle** [comment]
Ah right. Sorry - just getting back in to it, and a lot has been built and
changed - I will get there

**Comment by jimidle** [comment]
You can tell it isn’t the correct runtime because my changes require at
least go 1.18. See my comment in the PR on how you can use my version if
you want to try before Ter merges the PR.

By getting @dev, you are pulling in the old runtime.

Upgrade your go to go 1.19 as you will need it for the new runtime.

**Comment by parrt** [comment]
This one has been merged... Maybe you meant to comment on a different PR?

**Pull request merged**